File: 프로그래머스/lv4/64063. 호텔 방 배정/호텔 방 배정.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

unordered_map<long long, long long> m;

long long find_room(long long num)
{
    if (m[num] == 0) return num;
    else return m[num] = find_room(m[num]);
}
vector<long long> solution(long long k, vector<long long> room_number) {
    vector<long long> answer;

    for (auto num : room_number)
    {
        long long n = find_room(num);
        answer.push_back(n);
        m[n] = n+1;
    }
    return answer;
}

File: 프로그래머스/lv4/64063. 호텔 방 배정/호텔 방 배정.py
from collections import defaultdict
import sys
sys.setrecursionlimit(10000000)
m = defaultdict(int)

def find_room(n):
    if m[n] ==0:
        return n
    else:
        m[n] = find_room(m[n])
        return m[n]
def solution(k, room_number):
    answer = []
    for i in room_number:
        n = find_room(i)
        answer.append(n)
        m[n] = n+1
    return answer